controller/src/main/java/io/pravega/controller/eventProcessor/impl/EventProcessorGroupImpl.java

```diff
@@ -214,64 +242,50 @@ protected void shutDown() {
-            cell.stopAsync();
-        }
-        
-        for (EventProcessorCell<T> cell : eventProcessorMap.values()) {
-            try {
-                cell.awaitTerminated();
-                log.debug("Termination of event processor cell: {} completed successfully.", cell);
-            } catch (Exception e) {
-                log.warn("Failed terminating event processor cell {}.", cell, e);
-            }
-        }
-        
-        // Finally, clean up reader group from checkpoint store.
+    @Override
+    public void notifyProcessFailure(String controllerId) throws CheckpointStoreException {
+        long traceId = LoggerHelpers.traceEnterWithContext(log, this.objectId, "notifyProcessFailure", controllerId);
+        log.info("Notifying failure of process {} participating in reader group {}", controllerId, this.objectId);
         try {
-            log.debug("Attempting to clean up reader group {} entry from checkpoint store", this.objectId);
-            checkpointStore.removeProcessFromGroup(actorSystem.getProcess(), readerGroup.getGroupName());
-        } catch (CheckpointStoreException e) {
-            log.warn("Error removing reader group " + this.objectId, e);
-            return;
+            Map<String, Position> map = sealReaderGroup(controllerId);
+            cleanupReaderGroup(controllerId, map);
+        } finally {
+            LoggerHelpers.traceLeave(log, "notifyProcessFailure", traceId, controllerId);
         }
     }
 
-    @Override
-    public void notifyProcessFailure(String process) throws CheckpointStoreException {
-        long traceId = LoggerHelpers.traceEnterWithContext(log, this.objectId, "notifyProcessFailure", process);
-        log.info("Notifying failure of process {} participating in reader group {}", process, this.objectId);
+    private Map<String, Position> sealReaderGroup(String controllerId) throws CheckpointStoreException {
+        Map<String, Position> readerPositions = ImmutableMap.of();
         try {
-            Map<String, Position> map;
-            try {
-                log.debug("Removing reader group {} from process {}", readerGroup.getGroupName(), process);
-                map = checkpointStore.removeProcessFromGroup(process, readerGroup.getGroupName());
-            } catch (CheckpointStoreException e) {
-                if (e.getType().equals(CheckpointStoreException.Type.NoNode)) {
-                    return;
-                } else {
-                    throw e;
-                }
+            log.debug("Attempting to seal Reader Group {} for Controller {}", readerGroup.getGroupName(), controllerId);
+            // Change Reader Group State to SEALED so no new Readers can join
+            readerPositions = checkpointStore.sealReaderGroup(controllerId, readerGroup.getGroupName());
+        } catch (CheckpointStoreException e) {
+            if (!e.getType().equals(CheckpointStoreException.Type.NoNode)) {
+                throw e;
```

Comment:
The behavior has slightly changed here in the case of an exception different to not finding the zNode, right? In the previous code, in that case the method returns `null` whereas now it will return an empty map. I suppose that it does not change anything downstream, right? (e.g., a null check after this method)

Reply:
There is no change in behavior of `sealReaderGroup` : https://github.com/pravega/pravega/blob/f010db50b47a3461904a09506657fda0d34b8b19/controller/src/main/java/io/pravega/controller/store/checkpoint/ZKCheckpointStore.java#L127

Reply:
The refactored _if statement_  just does the same thing in a more concise way.....